### PR TITLE
Fix communicator issue

### PR DESCRIPTION
**Pull request by matt-frey**
For some program instances the MPI_COMM_WORLD MPI rank does not match the Cartesian MPI rank because the `reorder` flag does not seem to be taken into account by the MPI implementations despite the MPI specifications (see https://www.mpi-forum.org/docs/mpi-1.1/mpi-11-html/node133.html).

### CODE REVIEW

File: unit-tests/3d/test_mpi_parcel_communicate.f90 [hunk 1]
```diff
@@ -1,13 +1,13 @@
 ! =============================================================================
 !                         Test MPI parcel halo swap
 !
-!   This unit test checks parcel halo swap. Each MPI rank sends comm%rank+1
+!   This unit test checks parcel halo swap. Each MPI rank sends world%rank+1
```
Comment (sjboeing) [inline unit-tests/3d/test_mpi_parcel_communicate.f90:4]: Just checking these tests should be world related, not cart related.

File: unit-tests/3d/test_mpi_parcel_read.f90 [hunk 2]
```diff
@@ -19,25 +19,25 @@ program test_mpi_parcel_read
...
 
     !
     ! write parcels first
     !
 
-    n_parcels = 10 + (comm%rank + 1)
+    n_parcels = 10 + (world%rank + 1)
     n_parcels_before = n_parcels
-    n_total_parcels = 11 * comm%size + (comm%size * (comm%size - 1)) / 2
+    n_total_parcels = 11 * world%size + (world%size * (world%size - 1)) / 2
```
Comment (sjboeing) [inline unit-tests/3d/test_mpi_parcel_read.f90:34]: Here also, world vs cart based indexing.

File: unit-tests/3d/test_mpi_parcel_read_rejection.f90 [hunk 3]
```diff
@@ -94,13 +94,13 @@ program test_mpi_parcel_read_rejection
     y_sum = sum(parcels%position(2, 1:n_parcels))
     z_sum = sum(parcels%position(3, 1:n_parcels))
 
-    parcels%B(:, 1:n_parcels) = comm%rank + 1
-    parcels%volume(1:n_parcels) = comm%rank + 1
-    parcels%vorticity(:, 1:n_parcels) = comm%rank + 1
-    parcels%buoyancy(1:n_parcels) = comm%rank + 1
+    parcels%B(:, 1:n_parcels) = world%rank + 1
```
Comment (sjboeing) [inline unit-tests/3d/test_mpi_parcel_read_rejection.f90:97]: Same in this routine.

File: unit-tests/3d/test_mpi_parcel_split.f90 [hunk 4]
```diff
@@ -111,30 +111,30 @@ program test_mpi_parcel_split
     call mpi_blocking_reduce(n_orig_total, MPI_SUM)
 
     parcels%volume(1:n_parcels) = f12 * vcell
-    parcels%vorticity(:, 1:n_parcels) = comm%rank + 1
-    parcels%buoyancy(1:n_parcels) = comm%rank + 1
+    parcels%vorticity(:, 1:n_parcels) = world%rank + 1
```
Comment (sjboeing) [inline unit-tests/3d/test_mpi_parcel_split.f90:114]: And here.

File: unit-tests/3d/test_mpi_parcel_write.f90 [hunk 5]
```diff
@@ -8,31 +8,31 @@ program test_mpi_parcel_write
...
+    call mpi_env_initialise
 
-    passed = (passed .and. (comm%err == 0))
+    passed = (passed .and. (world%err == 0))
 
     call register_timer('parcel I/O', parcel_io_timer)
 
-    n_parcels = 10 + (comm%rank + 1)
-    n_total_parcels = 11 * comm%size + (comm%size * (comm%size - 1)) / 2
+    n_parcels = 10 + (world%rank + 1)
```
Comment (sjboeing) [inline unit-tests/3d/test_mpi_parcel_write.f90:23]: And here.

File: unit-tests/netcdf/test_mpi_netcdf_dataset_1d.f90 [hunk 6]
```diff
@@ -17,13 +17,13 @@ program test_mpi_netcdf_dataset_2d
     logical            :: passed = .true.
     integer            :: start(1), cnt(1)
 
-    call mpi_comm_initialise
+    call mpi_env_initialise
 
     do n = 1, n_local_parcels
-        dset(n) = n + n_local_parcels * comm%rank
+        dset(n) = n + n_local_parcels * world%rank
```
Comment (sjboeing) [inline unit-tests/netcdf/test_mpi_netcdf_dataset_1d.f90:23]: + here

File: unit-tests/netcdf/test_mpi_netcdf_dataset_2d.f90 [hunk 7]
```diff
@@ -16,10 +16,10 @@ program test_mpi_netcdf_dataset_2d
     logical            :: passed = .true.
     integer            :: xstart, xend, ystart, yend, xlen, ylen
 
-    call mpi_comm_initialise
+    call mpi_env_initialise
 
-    xstart = comm%rank * nx + 1
-    xend   = (comm%rank + 1) * nx
+    xstart = world%rank * nx + 1
```
Comment (sjboeing) [inline unit-tests/netcdf/test_mpi_netcdf_dataset_2d.f90:21]: world or cart?

File: unit-tests/netcdf/test_mpi_netcdf_dataset_3d.f90 [hunk 8]
```diff
@@ -44,7 +44,7 @@ program test_mpi_netcdf_dataset_3d
     do ix = lo(1), hi(1)
         do iy = lo(2), hi(2)
             do iz = lo(3), hi(3)
-                dset(iz, iy, ix) = comm%rank
+                dset(iz, iy, ix) = world%rank
```
Comment (sjboeing) [inline unit-tests/netcdf/test_mpi_netcdf_dataset_3d.f90:47]: Should this be world or cart?

File: unit-tests/netcdf/test_mpi_netcdf_read_dataset_1d.f90 [hunk 9]
```diff
@@ -19,17 +19,17 @@ program test_mpi_netcdf_read_dataset_1d
...
     allocate(wdset(n_local_parcels))
     allocate(rdset(n_local_parcels))
-    allocate(recvcounts(comm%size))
-    allocate(n_size(comm%size))
+    allocate(recvcounts(world%size))
+    allocate(n_size(world%size))
 
     do n = 1, n_local_parcels
-        wdset(n) = dble(n_local_parcels * comm%rank + n)
+        wdset(n) = dble(n_local_parcels * world%rank + n)
```
Comment (sjboeing) [inline unit-tests/netcdf/test_mpi_netcdf_read_dataset_1d.f90:32]: World or cart in this routine

File: unit-tests/netcdf/test_mpi_netcdf_read_dataset_2d.f90 [hunk 10]
```diff
@@ -17,10 +17,10 @@ program test_mpi_netcdf_read_dataset_2d
     logical                       :: passed = .true.
     integer                       :: xstart, xend, ystart, yend, xlen, ylen
 
-    call mpi_comm_initialise
+    call mpi_env_initialise
 
-    xstart = comm%rank * nx + 1
-    xend   = (comm%rank + 1) * nx
+    xstart = world%rank * nx + 1
```
Comment (sjboeing) [inline unit-tests/netcdf/test_mpi_netcdf_read_dataset_2d.f90:22]: World or cart in this routine?

**Review by sjboeing — COMMENTED** [review]
Great work, Matthias! I wasn't always sure whether world (rather than cart) was the right communicator. In these cases, I have put a note, but you may have figured these out already. Not sure if it is easy/necessary to run all the comm unit tests across 2 nodes to double-check all is OK.

File: unit-tests/3d/test_mpi_parcel_read.f90 (hunk 2 above)
Comment (matt-frey) [reply unit-tests/3d/test_mpi_parcel_read.f90:34]: `world%size = cart%size`

**Comment by matt-frey** [comment]
Indeed, I had to fix 3 unit-tests with respect to `world%rank` and `cart%rank`. Also, I forgot to initialise `cart%size`. This has now all been fixed. I ran the unit-tests with 6 cores distributed over 2 nodes (3 cores / node), 5 cores distributed over 2 nodes [layout 2-3], and 5 cores distributed over 3 nodes [layout 1-2-2].

File: unit-tests/3d/test_mpi_parcel_communicate.f90 (hunk 1 above)
Comment (matt-frey) [reply unit-tests/3d/test_mpi_parcel_communicate.f90]: Fixed.

File: unit-tests/3d/test_mpi_parcel_read_rejection.f90 (hunk 3 above)
Comment (matt-frey) [reply unit-tests/3d/test_mpi_parcel_read_rejection.f90:97]: Using `world%rank` is fine.

File: unit-tests/3d/test_mpi_parcel_split.f90 (hunk 4 above)
Comment (matt-frey) [reply unit-tests/3d/test_mpi_parcel_split.f90:114]: It is only initialised, but not used.

File: unit-tests/3d/test_mpi_parcel_write.f90 (hunk 5 above)
Comment (matt-frey) [reply unit-tests/3d/test_mpi_parcel_write.f90:23]: We could either use `cart%rank` or `world%rank`. It just has to be consistent throughout.

File: unit-tests/netcdf/test_mpi_netcdf_dataset_1d.f90 (hunk 6 above)
Comment (matt-frey) [reply unit-tests/netcdf/test_mpi_netcdf_dataset_1d.f90:23]: This is just some data. We could also use `cart%rank`. It does not really matter here.

File: unit-tests/netcdf/test_mpi_netcdf_dataset_2d.f90 (hunk 7 above)
Comment (matt-frey) [reply unit-tests/netcdf/test_mpi_netcdf_dataset_2d.f90:21]: Note, we do not set up the Cartesian communicator in this example at all. However, we could.

File: unit-tests/netcdf/test_mpi_netcdf_dataset_3d.f90 (hunk 8 above)
Comment (matt-frey) [reply unit-tests/netcdf/test_mpi_netcdf_dataset_3d.f90:47]: Note that this is just some arbitrary data. We can set the values to any number we like.

File: unit-tests/netcdf/test_mpi_netcdf_read_dataset_1d.f90 (hunk 9 above)
Comment (matt-frey) [reply unit-tests/netcdf/test_mpi_netcdf_read_dataset_1d.f90:32]: It does not matter. This is again just some data.

File: unit-tests/netcdf/test_mpi_netcdf_read_dataset_2d.f90 (hunk 10 above)
Comment (matt-frey) [reply unit-tests/netcdf/test_mpi_netcdf_read_dataset_2d.f90:22]: Note, we do not set up the Cartesian communicator in this example at all. However, we could.

**Pull request merged**